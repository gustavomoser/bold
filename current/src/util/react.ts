import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
