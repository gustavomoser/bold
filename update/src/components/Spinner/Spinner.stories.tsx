import { number } from '@storybook/addon-knobs'
import React from 'react'

import { Spinner } from './Spinner'

export default {
  title: 'Components/Spinner',
}

export const Default = () => <Spinner size={number('size', 1)} borderWidth={number('borderWidth', 2)} />
