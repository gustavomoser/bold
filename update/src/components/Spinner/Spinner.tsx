import React, { CSSProperties } from 'react'

import { ExternalStyles, Theme, useStyles } from '../../styles'

export interface SpinnerProps {
  size?: number
  borderWidth?: number
  style?: ExternalStyles
}

export function Spinner(props: SpinnerProps) {
  const { style } = props
  const { classes, css } = useStyles(createStyles, props)

  return <span className={css(classes.spinner, style)} />
}

Spinner.defaultProps = {
  size: 1,
  borderWidth: 2,
} as Partial<SpinnerProps>

export const createStyles = (theme: Theme, { size, borderWidth }: SpinnerProps) => ({
  spinner: {
    pointerEvents: 'none',
    fontSize: `${size}rem`,
    lineHeight: 1,
    ':after': {
      animation: `${theme.animation.spinAround} 450ms infinite linear`,
      border: `${borderWidth}px solid currentColor`,
      borderRadius: '50%',
      borderRightColor: 'transparent',
      content: '""',
      display: 'inline-block',
      height: `${size}rem`,
      width: `${size}rem`,
    },
  } as CSSProperties,
})
