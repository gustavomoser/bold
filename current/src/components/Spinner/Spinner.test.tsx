import { render } from '@testing-library/react'
import React from 'react'

import { Spinner } from './Spinner'

it('should render correctly', () => {
  const { container } = render(<Spinner />)
  expect(container).toMatchSnapshot()
})

it('should accept style prop', () => {
  const { container } = render(<Spinner style={{ color: 'red' }} />)
  expect(container).toMatchSnapshot()
})
