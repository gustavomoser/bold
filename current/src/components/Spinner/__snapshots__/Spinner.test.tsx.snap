// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept style prop 1`] = `
@keyframes animation-0 {
  from {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  to {
    -webkit-transform: rotate(359deg);
    -ms-transform: rotate(359deg);
    transform: rotate(359deg);
  }
}

.emotion-0 {
  pointer-events: none;
  font-size: 1rem;
  line-height: 1;
  color: red;
}

.emotion-0:after {
  -webkit-animation: animation-0 450ms infinite linear;
  animation: animation-0 450ms infinite linear;
  border: 2px solid currentColor;
  border-radius: 50%;
  border-right-color: transparent;
  content: "";
  display: inline-block;
  height: 1rem;
  width: 1rem;
}

<div>
  <span
    class="emotion-0"
  />
</div>
`;

exports[`should render correctly 1`] = `
@keyframes animation-0 {
  from {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }

  to {
    -webkit-transform: rotate(359deg);
    -ms-transform: rotate(359deg);
    transform: rotate(359deg);
  }
}

.emotion-0 {
  pointer-events: none;
  font-size: 1rem;
  line-height: 1;
}

.emotion-0:after {
  -webkit-animation: animation-0 450ms infinite linear;
  animation: animation-0 450ms infinite linear;
  border: 2px solid currentColor;
  border-radius: 50%;
  border-right-color: transparent;
  content: "";
  display: inline-block;
  height: 1rem;
  width: 1rem;
}

<div>
  <span
    class="emotion-0"
  />
</div>
`;
